src/Illuminate/Database/Eloquent/SoftDeletes.php

```diff
@@ -69,35 +72,52 @@ protected function performDeleteOnModel()
             return $this->setKeysForSaveQuery($this->newModelQuery())->forceDelete();
         }
 
-        return $this->runSoftDelete();
+        $this->runSoftDelete($this->freshTimestamp(), 'trashed');
     }
 
     /**
      * Perform the actual delete query on this model instance.
      *
+     * @param  \DatetimeInterface|string  $time
+     * @param  string  $event
      * @return void
      */
-    protected function runSoftDelete()
+    protected function runSoftDelete($time, $event)
     {
         $query = $this->setKeysForSaveQuery($this->newModelQuery());
 
-        $time = $this->freshTimestamp();
-
         $columns = [$this->getDeletedAtColumn() => $this->fromDateTime($time)];
 
         $this->{$this->getDeletedAtColumn()} = $time;
 
         if ($this->timestamps && ! is_null($this->getUpdatedAtColumn())) {
-            $this->{$this->getUpdatedAtColumn()} = $time;
+            $updatedAt = $this->freshTimestamp();
 
-            $columns[$this->getUpdatedAtColumn()] = $this->fromDateTime($time);
+            $this->{$this->getUpdatedAtColumn()} = $updatedAt;
+
+            $columns[$this->getUpdatedAtColumn()] = $this->fromDateTime($updatedAt);
         }
 
         $query->update($columns);
 
         $this->syncOriginalAttributes(array_keys($columns));
 
-        $this->fireModelEvent('trashed', false);
+        $this->fireModelEvent($event, false);
+    }
+
+    /**
+     * Sets the model to be deleted in the future.
+     *
+     * @param  \DateTimeInterface|string  $datetime
+     * @return void
+     */
+    public function trashAt($datetime)
+    {
+        if ($this->freshTimestamp() >= $datetime) {
```

Comment:
I think there might be a potential issue here.

If I do something like

```php
$now = Date::now();
$model->trashAt($now);
```

then by the time we get to `freshTimestamp()` there might be a tick resulting in the return value of that being greater than the provided datetime.  

I wonder if it might be okay to avoid this check altogether and allow people to delete something in the past?  Maybe I'm backfilling bad data or something?  Alternatively, you could just grab the greater of `$this->freshTimestamp()` and `$datetime`.  Though that probably would be surprising behavior for many so I don't love that idea.

Reply:
Well in this case, the user should use `delete` or `trash` function if he wants to delete the model instantly, the `trashAt` is used to "trash" in the future.
This is my understanding.

Reply:
Now that you mention it, a millisecond difference could force the developer to force it around by adding more minutes.

Seems that allowing to trash in the past could be the solution. The problem is what event will fire. 

Theorically, a condition on the timestamp at firing time should be fine.

Reply:
> I think there might be a potential issue here.
> 
> 
> 
> If I do something like
> 
> 
> 
> ```php
> 
> $now = Date::now();
> 
> $model->trashAt($now);
> 
> ```
> 
> 
> 
> then by the time we get to `freshTimestamp()` there might be a tick resulting in the return value of that being greater than the provided datetime.  
> 
> 
> 
> I wonder if it might be okay to avoid this check altogether and allow people to delete something in the past?  Maybe I'm backfilling bad data or something?  Alternatively, you could just grab the greater of `$this->freshTimestamp()` and `$datetime`.  Though that probably would be surprising behavior for many so I don't love that idea.

You use `delete()` to delete something now.